applications/machine_learning/src/modules/ml_runner.c

```diff
@@ -48,20 +54,97 @@ static void result_ready_cb(int err)
 
 	if (err) {
 		LOG_ERR("Cannot get classification results (err: %d)", err);
+		module_set_state(MODULE_STATE_ERROR);
 		k_free(evt);
-		report_error();
 	} else {
 		EVENT_SUBMIT(evt);
+	}
+}
+
+static void report_error(void)
+{
+	state = STATE_ERROR;
+	module_set_state(MODULE_STATE_ERROR);
+}
 
-		err = ei_wrapper_start_prediction(SHIFT_WINDOWS, SHIFT_FRAMES);
+static int buf_cleanup(void)
+{
+	bool cancelled;
+	int err = ei_wrapper_clear_data(&cancelled);
 
-		if (err) {
-			LOG_ERR("Cannot restart prediction (err: %d)", err);
-			report_error();
+	if (!err) {
+		if (cancelled) {
+			ml_control &= ~ML_RUNNING;
 		}
+
+		if (ml_control & ML_RUNNING) {
+			ml_control |= ML_DROP_RESULT;
+		}
+
+		ml_control &= ~ML_CLEANUP_REQUIRED;
+		ml_control |= ML_FIRST_PREDICTION;
+	} else if (err == -EBUSY) {
+		__ASSERT_NO_MSG(ml_control & ML_RUNNING);
+		ml_control |= ML_DROP_RESULT;
+		ml_control |= ML_CLEANUP_REQUIRED;
+	} else {
+		LOG_ERR("Cannot cleanup buffer (err: %d)", err);
+		report_error();
+	}
+
+	return err;
+}
+
+static void start_prediction(void)
+{
+	int err;
+	size_t window_shift;
+	size_t frame_shift;
+
+	if (ml_control & ML_RUNNING) {
+		return;
+	}
+
+	if (ml_control & ML_CLEANUP_REQUIRED) {
+		if (buf_cleanup()) {
+			return;
+		}
+	}
+
+	if (ml_control & ML_FIRST_PREDICTION) {
+		window_shift = 0;
+		frame_shift = 0;
+	} else {
+		window_shift = SHIFT_WINDOWS;
+		frame_shift = SHIFT_FRAMES;
+	}
+
+	err = ei_wrapper_start_prediction(window_shift, frame_shift);
```

Comment:
`int err`

Reply:
Used to store results of buf_cleanup a few lines above